packages/flutter_tools/lib/src/fuchsia/fuchsia_device.dart

```diff
@@ -621,40 +656,82 @@ class FuchsiaDevice extends Device {
 
   String _cachedHostAddress;
 
-  /// List the ports currently running a dart observatory.
-  Future<List<int>> servicePorts() async {
-    const String findCommand = 'find /hub -name vmservice-port';
+  /// List the ports currently running a dart observatory and the component url
+  /// of flutter views that such dart observatory provides
+  Future<Map<int, List<String>>> servicePorts() async {
+    // TODO(michaellee8): Use the v2 /hub protocol,
+    //  https://github.com/flutter/flutter/issues/83609#issuecomment-871114230,
+    //  https://bugs.fuchsia.dev/p/fuchsia/issues/detail?id=79924,
+    //  https://fuchsia.dev/fuchsia-src/concepts/components/v2/hub
+
+    final Iterable<String> servicePaths = await vmServicePaths();
+
+    final Map<int, List<String>> portUrlMap = <int, List<String>>{};
+
+    // In practice it should be safe to assume that there are only one port
+    // for each vmservice-port, but to play safe we duplicates entries if there
+    // are multiple ports for the path.
+    for (final String path in servicePaths) {
+
+      // Get ports
+      final Iterable<int> ports = (await listDirectory(path)).map(int.tryParse);
+
+      // Get component urls
+      final List<String> urls = await
+          Stream<String>.fromIterable(await findByName('$path/../../../c', 'url'))
+          .asyncMap(readFile).toList();
+
+      for (final int port in ports){
+        portUrlMap[port] = urls;
+      }
+    }
+    return portUrlMap;
+  }
+
+  @visibleForTesting
+  Future<Iterable<String>> vmServicePaths() async {
+    final Iterable<String> paths = await findByName('/hub', 'vmservice-port');
```

Comment:
This can be extremely slow and fragile. It is preferable to use the inspect apis to query for the port.

Reply:
something along the lines of `iquery -f json show root:vm_service_port --manifest flutter_jit_runner.cmx` will give you the port in a json payload.

Reply:
or if you want to cover both aot and jit you can do `iquery -f json show "flutter_*_runner.cmx:root:vm_service_port"`.

Reply:
Let me try to use the inspect api.

Reply:
@chaselatta Do the flutter_tools codebase uses any kind of more structured json parsing library like json_serializable? It is pretty hard to do json parsing on iquery result without one.

Reply:
The tool uses the json parser in `dart:convert`. See https://api.dart.dev/dev/2.14.0-305.0.dev/dart-convert/json-constant.html

Reply:
@zanderso So doing some parsing by hand is unavoidable anyway?

Reply:
@michaellee8 I don't know what the level of complexity of the json coming from iquery is, but if you only need to extract a few things, then doing that manually is preferable to introducing a new dependency to the tool.

Reply:
@zanderso I still find trying to parse those json returned by iquery is really inconvenient, so I wrote a small helper for it to do it in tree, in https://github.com/flutter/flutter/pull/86620. Please help to review it and decide if it can be accepted. I will do the coding of parsing iquery using it if it is accepted.

Reply:
@michaellee8 We are currently landing a change to put the VM service port publishing on the component instead of the runner. The CL which is adding this is https://fuchsia-review.googlesource.com/c/fuchsia/+/606062. This means that you will be able to launch your component and query the component for the port instead of having to query the runner. This is part of a larger effort to clean up service port discovery on flutter devices. 

Another approach would be to not do this manually but instead use ffx flutter tunnel to set up the tunnels for you.